optimism/material/HyperViscoelastic.py

```diff
@@ -0,0 +1,216 @@
+    #     print('  Relaxation time = %s' % properties['relaxation time %s' % n])
+    #     props = np.hstack(
+    #         (props, np.array([properties['non equilibrium shear modulus %s' % n],
+    #                           properties['relaxation time %s' % n]])))
+
+    
+
+    return props
+
+def _energy_density(dispGrad, state, dt, props):
+    W_eq  = _eq_strain_energy(dispGrad, props)
+    W_neq = _neq_strain_energy(dispGrad, state, dt, props)
+    return W_eq + W_neq
+
+# TODO generalize to arbitrary strain energy density
+def _eq_strain_energy(dispGrad, props):
+    K, G = props[PROPS_K_eq], props[PROPS_G_eq]
+    F = dispGrad + np.eye(3)
+    J = np.linalg.det(F)
+    J23 = np.power(J, -2.0 / 3.0)
+    I1Bar = J23 * np.tensordot(F,F)
+    Wvol = 0.5 * K * (0.5 * J**2 - 0.5 - np.log(J))
+    Wdev = 0.5 * G * (I1Bar - 3.0)
+    return Wdev + Wvol
+
+def _neq_strain_energy(dispGrad, stateOld, dt, props):
+    I = np.identity(3)
+    F = dispGrad + I
+    state_new = _compute_state_new(dispGrad, stateOld, dt, props)
+    # Fvs = state_new.reshape((NUM_PRONY_TERMS, 3, 3))
+    Fv_new = state_new.reshape((3, 3))
+
+    G_neq = props[PROPS_G_neq]
+    tau   = props[PROPS_TAU]
+    eta   = G_neq * tau
+
+    Fe = F @ np.linalg.inv(Fv_new)
+    Ee = 0.5 * TensorMath.mtk_log_sqrt(Fe.T @ Fe)
+    Me = 2. * G_neq * Ee
+    M_bar = TensorMath.norm_of_deviator_squared(Me)
+    gamma_dot = M_bar / eta
+    # visco_energy = (dt / (G_neq * tau)) * M_bar**2
+    visco_energy = 0.5 * dt * eta * gamma_dot**2
+
+    W_neq = G_neq * TensorMath.norm_of_deviator_squared(Ee) + visco_energy
+    # def vmap_body(n, Fv):
+    #     G_neq = props[PROPS_G_neq + 2 * n]
+    #     # tau   = props[PROPS_TAU   + 2 * n]
+
+    #     Fe = F @ np.linalg.inv(Fv)
+    #     Ee = 0.5 * TensorMath.mtk_log_sqrt(Fe.T @ Fe)
+
+    #     # viscous shearing
+    #     # Me = 2. * G_neq * Ee
+    #     # M_bar = TensorMath.norm_of_deviator_squared(Me)
+    #     # visco_energy = (dt / (G_neq * tau)) * M_bar**2
+
+    #     # still need another term I think
+    #     W_neq = G_neq * TensorMath.norm_of_deviator_squared(Ee) #+ visco_energy
+    #     return W_neq
+
+    # W_neq = np.sum(vmap(vmap_body, in_axes=(0, 0))(np.arange(NUM_PRONY_TERMS), Fvs))
+
+    return W_neq
+
+# state update
+def _compute_state_new(dispGrad, stateOld, dt, props):
+    state_inc = _compute_state_increment(dispGrad, stateOld, dt, props)
+    # Fv_olds   = stateOld.reshape((NUM_PRONY_TERMS, 3, 3))
+    # Fv_incs   = state_inc.reshape((NUM_PRONY_TERMS, 3, 3))
+
+    # def vmap_body(n, Fv_old, Fv_inc):
+    #     Fv_new = Fv_inc @ Fv_old
+    #     return Fv_new.ravel()
+
+    # state_new = np.hstack(vmap(vmap_body, in_axes=(0, 0, 0))(np.arange(NUM_PRONY_TERMS), Fv_olds, Fv_incs))
+
+    Fv_old = stateOld.reshape((3, 3))
+    Fv_inc = state_inc.reshape((3, 3))
+    state_new = (Fv_inc @ Fv_old).ravel()
+    return state_new
+
+def _compute_state_increment(dispGrad, stateOld, dt, props):
```

Comment:
The J2 model had this function so that we could put the finite strain and the infinitesimal strain versions in one implementation. For this model, which only makes sense as a finite deformation model, we might as well do all of the update in the `_compute_state_new()` function which will make it easier to read. Maybe this will speed things up a little, too.